src/main/java/com/scalar/cassy/transferer/AzureBlobFileDownloader.java

```diff
@@ -27,33 +29,33 @@ public AzureBlobFileDownloader(BlobContainerAsyncClient blobContainerClient) {
   @Override
   public void download(RestoreConfig config) {
     String key = BackupPath.create(config, config.getKeyspace());
-    List<CompletableFuture> futures = new ArrayList<>();
-    blobContainerClient
-        .listBlobs(new ListBlobsOptions().setPrefix(key))
-        .toIterable()
-        .forEach(
-            b -> {
-              try {
-                Files.createDirectories(Paths.get(config.getDataDir(), b.getName()).getParent());
-              } catch (IOException e) {
-                throw new FileTransferException(e);
-              }
-              futures.add(
-                  blobContainerClient
-                      .getBlobAsyncClient(b.getName())
-                      .downloadToFile(Paths.get(config.getDataDir(), b.getName()).toString())
-                      .toFuture());
-            });
 
-    logger.info("Downloading " + blobContainerClient.getBlobContainerName() + key);
-    futures.forEach(
-        f -> {
-          try {
-            f.get();
-          } catch (InterruptedException | ExecutionException e) {
-            throw new FileTransferException(e);
-          }
-        });
+    logger.info("Downloading " + blobContainerClient.getBlobContainerName() + "/" + key);
+    List<Mono<BlobProperties>> filesToBeDownloaded = new ArrayList<>();
+    Iterable<BlobItem> keyspaceBlobs =
+        blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/")).toIterable();
+    for (BlobItem blob : keyspaceBlobs) {
+      Path destFile = Paths.get(config.getDataDir(), blob.getName());
+      try {
+        Files.createDirectories(destFile.getParent());
+      } catch (IOException e) {
+        throw new FileTransferException(e);
+      }
+      filesToBeDownloaded.add(
+          blobContainerClient
+              .getBlobAsyncClient(blob.getName())
+              .downloadToFile(destFile.toString())
+              .doOnSuccess(
+                  blobProperties -> logger.info("Download file succeeded : " + destFile.toString()))
+              .doOnError(
+                  error -> {
+                    throw new FileTransferException(
+                        "Download file failed : " + destFile.toString(), error);
+                  }));
+    }
+
+    // Start asynchronously all downloads and wait for all of them to complete
+    Mono.when(filesToBeDownloaded).block();
```

Comment:
@feeblefakie 
Compared to your last review, when an error occurs, the `doOnError` will catch and rethrow a `FileTransferException` which means the try/catch block around the `Mono.when(...)` is not necessary anymore.